File: packages/shares/src/api/author.ts
import { del, get, patch, post } from '../utils';
import type {
  Book,
  BookDraft,
  Catalog,
  CatalogDraft,
} from '@guga-reading/types';
export const useApiAuthor = {
  prefix: '/author',
  getBook: async function (id: number = -1) {
    return await get<Book[]>(
      `${this.prefix}/book?${id != -1 ? `id=${id}` : ''}`,
    );
  },
  /**
   * 获取书籍目录, 如果chapterId为-1则返回所有章节
   * @param bookId   书籍id
   * @param chapterId   章节id
   * @returns
   */
  getBookCatalog: async function (bookId: number, chapterId: number = -1) {
    return await get<Catalog[]>(
      `${this.prefix}/book-catalog?book_id=${bookId}&chapter_id=${chapterId}`,
    );
  },
  /**
   * 删除书籍章节
   * @param bookId  书籍id
   * @returns
   */
  deleteBookChapter: async function (
    bookId: number,
    sort_orders: number[],
    is_draft: boolean,
  ) {
    return await del<null>(
      `${this.prefix}/book-chapter/${bookId}?is_draft=${is_draft}`,
      {
        data: {
          sort_orders,
        },
      },
    );
  },
  /**
   * 获取书籍章节草稿,如果chapterId为-1则返回所有章节
   * @param bookId  书籍id
   * @param chapterId   章节id
   * @returns
   */
  getBookChapterDraft: async function (bookId: number, chapterId: number = -1) {
    return await get<CatalogDraft[]>(
      `${this.prefix}/book-chapter-draft?book_id=${bookId}&chapter_id=${chapterId}`,
    );
  },
  /**
   * 获取书籍章节
   * @param bookId  书籍id
   * @param sortOrder  章节排序
   * @param isDraft  是否草稿
   * @returns
   */
  getBookChapter: async function (
    bookId: number,
    sortOrder: number,
    isDraft: boolean,
  ) {
    return await get<string>(
      `${this.prefix}/book-chapter?book_id=${bookId}&sort_order=${sortOrder}&is_draft=${isDraft}`,
    );
  },
  /**
   * 获取书籍章节草稿
   * @param bookId  书籍id
   * @param sortOrder  章节排序
   * @returns
   */
  getBookChapterDraftItem: async function (bookId: number, sortOrder: number) {
    return await get<CatalogDraft>(
      `${this.prefix}/book-chapter-draft-item?book_id=${bookId}&sort_order=${sortOrder}`,
    );
  },
  /**
   * 保存书籍章节
   * @param bookId  书籍id
   * @param sortOrder  章节排序
   * @param isDraft  是否草稿
   * @param content  章节内容
   * @param title  章节标题
   * @returns
   */
  saveBookChapter: async function (
    bookId: number,
    sortOrder: number,
    isDraft: boolean,
    content: string,
    title: string,
  ) {
    return await patch<null>(`${this.prefix}/chapter`, {
      book_id: bookId,
      sort_order: sortOrder,
      is_draft: isDraft,
      content,
      title,
    });
  },
  /**
   * 创建书籍章节
   * @param bookId  书籍id
   * @param title  章节标题
   * @param content  章节内容
   * @param sortOrder  章节排序
   * @returns
   */
  createBookChapter: async function (
    bookId: number,
    title: string,
    content: string,
    sortOrder: number,
  ) {
    return await post<null>(`${this.prefix}/chapter`, {
      book_id: bookId,
      title,
      content,
      sort_order: sortOrder,
    });
  },
  /**
   * 创建书籍
   * @param name  书籍名称
   * @param author  作者
   * @param cover  封面
   * @param description  描述
   * @param category  类别
   * @param tags  标签
   * @returns
   */
  createBook: async function (
    name: string,
    author: string,
    cover: File,
    description: string,
    category: string,
    tags: string,
  ) {
    const formData = new FormData();
    formData.append('name', name);
    formData.append('author', author);
    formData.append('cover', cover);
    formData.append('description', description);
    formData.append('category', category);
    formData.append('tags', tags);
    return await post<null>(`${this.prefix}/book`, formData);
  },
  /**
   * 提交书籍章节
   * @param bookId  书籍id
   * @param sortOrder  章节排序
   * @returns
   */
  submitBookChapter: async function (bookId: number, sortOrder: number) {
    return await patch<null>(
      `${this.prefix}/status/book-chapter?book_id=${bookId}&sort_order=${sortOrder}`,
    );
  },
  /**
   * 获取书籍草稿
   * @param id  书籍草稿id
   * @returns
   */
  getBookDraft: async function (id: number = -1) {
    return await get<BookDraft[]>(`${this.prefix}/book-draft?id=${id}`);
  },
  /**
   * 提交书籍草稿
   * @returns
   */
  submitBookDraft: async function (bookDraftId: number) {
    return await patch<null>(
      `${this.prefix}/status/book-draft?id=${bookDraftId}`,
    );
  },
  /**
   * 删除书籍草稿
   * @param bookDraftId  书籍草稿id
   * @param action  删除操作
   * @returns
   */
  deleteBookDraft: async function (bookDraftId: number, action: string) {
    return await del<null>(
      `${this.prefix}/book-draft?id=${bookDraftId}&action=${action}`,
    );
  },
  updateBook: async function (
    id: number,
    name: string,
    author: string,
    cover: File,
    description: string,
    category: string,
    tags: string,
    isDraft: boolean,
    action: string,
  ) {
    const formData = new FormData();
    formData.append('id', id.toString());
    formData.append('name', name);
    formData.append('author', author);
    if (cover) {
      formData.append('cover', cover);
    }
    formData.append('description', description);
    formData.append('category', category);
    formData.append('tags', tags);
    formData.append('is_draft', isDraft.toString());
    formData.append('action', action);
    return await patch<null>(`${this.prefix}/book`, formData);
  },
};

File: ktorBackend/src/main/kotlin/com/qianrenni/controller/author.kt
package com.qianrenni.controller

import com.qianrenni.enums.ActionEnum
import com.qianrenni.enums.ResourceTypeEnum
import com.qianrenni.enums.ScopeEnum
import com.qianrenni.plugins.PermissionCheck
import com.qianrenni.plugins.getCurrentUser
import com.qianrenni.schemas.ResponseModel
import com.qianrenni.services.authorBookService
import com.qianrenni.services.generatePermissionCode
import io.ktor.http.content.*
import io.ktor.server.auth.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import io.ktor.util.cio.*
import io.ktor.utils.io.*
import kotlinx.serialization.Serializable
import java.io.File
import kotlin.io.path.createTempFile

@Serializable
data class RequestUpdateBookChapter(
    val bookId: Int,
    val title: String,
    val order: Float,
    val content: String
)
fun Routing.author() {
    route("/author") {
        authenticate("auth-jwt") {
            // GET /author/count - 获取作者数量
            install(PermissionCheck) {
                requiredPermissions = listOf(
                    generatePermissionCode(
                        resource = ResourceTypeEnum.BOOK, action = ActionEnum.CREATE, scope = ScopeEnum.OWN
                    )
                )
            }
            get("/count") {
                val result = application.authorBookService.getAuthorCount()
                call.respond(ResponseModel.Success(result))
            }

            // GET /author/book - 获取作者图书列表
            get("/book") {
                val user = call.getCurrentUser()
                val bookIds = call.request.queryParameters.getAll("id")?.map { it.toInt() } ?: emptyList()
                val result = application.authorBookService.getBook(userId = user.id, bookIds = bookIds)
                call.respond(ResponseModel.Success(result))
            }

            // POST /author/book - 创建作者图书
            post("/book") {
                val user = call.getCurrentUser()
                val multipartData = call.receiveMultipart(formFieldLimit = 1024 * 1024)
                var bookName: String? = null
                var author: String? = null
                var description: String? = null
                var category: String? = null
                var tags: String? = null
                var cover: File? = null
                multipartData.forEachPart { part ->
                    when (part) {
                        is PartData.FormItem -> {
                            when (part.name) {
                                "name" -> {
                                    bookName = part.value
                                }

                                "author" -> {
                                    author = part.value
                                }

                                "description" -> {
                                    description = part.value
                                }

                                "category" -> {
                                    category = part.value
                                }

                                "tags" -> {
                                    tags = part.value
                                }
                            }
                        }

                        is PartData.FileItem -> {
                            val fileName = part.originalFileName
                            fileName?.let {
                                cover = createTempFile("cover_${user.id}", it.split(".").last()).toFile()
                            }
                            cover?.let {
                                part.provider().copyAndClose(it.writeChannel())
                            }
                        }

                        else -> {}
                    }
                    part.release()
                }
                //全部不为空才创建
                if (
                    bookName != null
                    && author != null
                    && description != null
                    && category != null
                    && tags != null
                    && cover != null
                ) {
                    application.authorBookService.createBook(
                        userId = user.id,
                        bookName = bookName,
                        author = author,
                        tags = tags,
                        description = description,
                        category = category,
                        coverFile = cover
                    )
                }
                call.respond(ResponseModel.Empty(message = "创建成功"))
            }

            // PATCH /author/book - 更新作者图书
            patch("/book") {
                val user = call.getCurrentUser()
                val multipartData = call.receiveMultipart(formFieldLimit = 1024 * 1024)
                var bookName: String? = null
                var author: String? = null
                var description: String? = null
                var category: String? = null
                var tags: String? = null
                var cover: File? = null
                var bookId: Int? = null
                multipartData.forEachPart { part ->
                    when (part) {
                        is PartData.FormItem -> {
                            when (part.name) {
                                "id" -> {
                                    bookId = part.value.toInt()
                                }

                                "name" -> {
                                    bookName = part.value
                                }

                                "author" -> {
                                    author = part.value
                                }

                                "description" -> {
                                    description = part.value
                                }

                                "category" -> {
                                    category = part.value
                                }

                                "tags" -> {
                                    tags = part.value
                                }
                            }
                        }

                        is PartData.FileItem -> {
                            val fileName = part.originalFileName
                            fileName?.let {
                                cover = createTempFile("cover_${user.id}", it.split(".").last()).toFile()
                            }
                            cover?.let {
                                part.provider().copyAndClose(it.writeChannel())
                            }
                        }

                        else -> {}
                    }
                    part.release()
                }
                //全部不为空才创建
                if (
                    bookId != null
                    && bookName != null
                    && author != null
                    && description != null
                    && category != null
                    && tags != null
                ) {
                    application.authorBookService.updateBook(
                        userId = user.id,
                        bookId = bookId,
                        bookName = bookName,
                        author = author,
                        tags = tags,
                        description = description,
                        category = category,
                        coverFile = cover
                    )
                }
                call.respond(ResponseModel.Empty(message = "更新成功"))
            }

            // DELETE /author/book - 删除作者图书
            delete("/book") {
                val user = call.getCurrentUser()
                val bookId = call.requireQueryParameter("id").toInt()
                application.authorBookService.deleteBook(userId = user.id, bookId = bookId)
                call.respond(ResponseModel.Empty(message = "删除成功"))
            }

            // GET /author/chapter - 获取作者图书章节
            get("/chapter") {
                val user = call.getCurrentUser()
                val bookId = call.requireQueryParameter("bookId").toInt()
                val chapterId = call.request.queryParameters.getAll("chapterId")?.map { it.toInt() }
                val result = application.authorBookService.getBookChapter(
                    userId = user.id,
                    bookId = bookId,
                    chapterId = chapterId
                )
                call.respond(ResponseModel.Success(result))
            }

            // PATCH /author/chapter - 更新作者图书章节
            patch("/chapter") {
                val user = call.getCurrentUser()
                val requestUpdateBookChapter = call.receive<RequestUpdateBookChapter>()
                application.authorBookService.updateBookChapter(
                    userId = user.id,
                    requestUpdateBookChapter = requestUpdateBookChapter
                )
                call.respond(ResponseModel.Empty(message = "更新成功"))
            }

            // DELETE /author/chapter - 删除作者图书章节
            delete("/chapter") {
                val user = call.getCurrentUser()
                val chapterId = call.requireQueryParameter("chapterId").toInt()
                val bookId = call.requireQueryParameter("bookId").toInt()
                application.authorBookService.deleteBookChapter(
                    userId = user.id,
                    chapterId = chapterId,
                    bookId = bookId
                )
                call.respond(ResponseModel.Empty(message = "删除成功"))
            }

            // GET /author/book-statistics - 获取作者图书阅读数据
            get("/book-statistics") {
                val user = call.getCurrentUser()
                val bookId = call.requireQueryParameter("bookId").toInt()
                val result = application.authorBookService.getBookReadStatistic(userId = user.id, bookId = bookId)
                call.respond(ResponseModel.Success(result))
            }

            // GET /author/content - 获取章节内容
            get("/content") {
                val user = call.getCurrentUser()
                val chapterIds = call.request.queryParameters.getAll("chapterId")?.map { it.toInt() } ?: emptyList()
                val bookId = call.requireQueryParameter("bookId").toInt()
                val result = application.authorBookService.getChapterContent(
                    userId = user.id,
                    chapterIds = chapterIds,
                    bookId = bookId
                )
                call.respond(ResponseModel.Success(result))
            }

            // GET /author/draft/chapter - 获取草稿章节
            route("/draft") {
                get("/chapter") {
                    val user = call.getCurrentUser()
                    val result = application.authorBookService.getDraftChapter(userId = user.id)
                    call.respond(ResponseModel.Success(result))
                }
            }
            route("/status") {
                // PATCH /author/status/chapter - 更新章节状态
                patch("chapter") {
                    val user = call.getCurrentUser()
                    val bookId = call.requireQueryParameter("bookId").toInt()
                    val chapterId = call.requireQueryParameter("chapterId").toInt()
                    application.authorBookService.updateStatusChapter(
                        userId = user.id,
                        bookId = bookId,
                        chapterId = chapterId
                    )
                    call.respond(ResponseModel.Empty(message = "更新成功"))
                }

                // PATCH /author/status/book - 更新书籍状态
                patch("/book") {
                    val user = call.getCurrentUser()
                    val bookId = call.requireQueryParameter("bookId").toInt()
                    application.authorBookService.updateStatusBook(userId = user.id, bookId = bookId)
                    call.respond(ResponseModel.Empty(message = "更新成功"))
                }
            }

        }
    }
}
